lib/ansible/galaxy/collection/__init__.py

```diff
@@ -681,29 +407,125 @@ def install_collections(collections, output_path, apis, validate_certs, ignore_e
     :param force_deps: Re-install a collection as well as its dependencies if they have already been installed.
     """
-    existing_collections = find_existing_collections(output_path, fallback_metadata=True)
+    existing_collections = {
+        Requirement(coll.fqcn, coll.ver, coll.src, coll.type)
+        for coll in find_existing_collections(output_path, artifacts_manager)
+    }
 
-    with _tempdir() as b_temp_path:
-        display.display("Process install dependency map")
-        with _display_progress():
-            dependency_map = _build_dependency_map(collections, existing_collections, b_temp_path, apis,
-                                                   validate_certs, force, force_deps, no_deps,
-                                                   allow_pre_release=allow_pre_release)
-
-        display.display("Starting collection install process")
-        with _display_progress():
-            for collection in dependency_map.values():
-                try:
-                    collection.install(output_path, b_temp_path)
-                except AnsibleError as err:
-                    if ignore_errors:
-                        display.warning("Failed to install collection %s but skipping due to --ignore-errors being set. "
-                                        "Error: %s" % (to_text(collection), to_text(err)))
-                    else:
-                        raise
+    if force_deps:
+        breakpoint()
+
+    requested_requirements = set(
+        chain.from_iterable(
+            (
+                Requirement.from_dir_path(sub_coll, artifacts_manager)
+                for sub_coll in (
+                    artifacts_manager.
+                    get_direct_collection_dependencies(install_req).
+                    keys()
+                )
+            )
+            if install_req.is_subdirs else (install_req, )
+            for install_req in collections
+        ),
+    )
+    requested_requirements_names = {req.fqcn for req in requested_requirements}
+
+    # NOTE: Don't attempt to reevaluate already installed deps
+    # NOTE: unless `--force` is passed
+    # FIXME: debug why `--force` doesn't always work well
+    requested_requirements -= set() if force else {
+        req
+        for req in requested_requirements
+        for exs in existing_collections
+        if req.fqcn == exs.fqcn and meets_requirements(exs.ver, req.ver)
+    }
 
+    # FIXME: This probably needs to be improved to 
+    # FIXME: properly match differing src/type.
+    existing_non_requested_collections = {
+        coll for coll in existing_collections
+        if coll.fqcn not in requested_requirements_names
+    }
+    # NOTE: Pin installed collection versions that are
+    # NOTE: not directly requested via CLI if `--force-deps` is not set
+    requested_requirements |= (
+        set() if force_deps
+        else existing_non_requested_collections
+    )
+
+    if not requested_requirements:
+        display.display(
```

Comment:
I'm fine with the change in message, just noting that this test needs to be fixed up to check the new assertion https://github.com/ansible/ansible/blob/cf39d9de258cb9c47de9043e1a85e327e177dba7/test/integration/targets/ansible-galaxy-collection/tasks/install.yml#L43.